### ISSUE TITLE
Cascading encryption

### ISSUE THREAD

**Issue by hicksel** [irungentoo/toxcore#1209]
As far as I can tell from the info on http://nacl.cr.yp.to/sign.html crypto_sign_edwards25519sha512batch is used and consists of Curve25519 as encryption algorithm and SHA-512 as hash algorithm. But as Curve25519 is pretty new and SHA-512 is thought to be safe, but an American standard I think adding the option of cascading encryption and hash algorithms into toxcore would greatly improve security, especially if messages are supposed to sustain future attacks.
Therefore I would suggest adding RIPDEM-320(the longest version of RIPEMD-160), which a version of was also used in Bitcoin(Bitcoin uses SHA-256 and RIPEMD-160) as default second hash algorithm which would always be used. Also there should be an option to even add a third hash algorithm like whirlpool if both clients support it.
To improve Security even more I suggest adding RSA on top of the Curve25519 encryption as following:
As soon as two friends (Bob and Alice) have added each other and have checked the other´s public key, Bob who wants to write a message to Alice whilst using a client that by default wants to use multiple encryption algorithms. So for this case the toxcore would need to be extended, so that other encryption algorithms can be applied on top of the Curve25519 encryption.

Initial Steps (on Bob’s end, who wants to enhance security)
1. Bob: toxcore wants to add a 2nd layer of encryption (because of user preset/decision of the tox developers to use it by default)
2. Bob: if not already done toxcore generates a new Public Key/Private Key Keypair suitable for the new encryption algorithm (for example RSA-4096)

Exchanging Keys with current friends (has to be done for every friend of Bob)
Key Exchange:
1. Bob: toxcore sends a message to Alice with the encryption algorithm Bob wants to add and his Public Key for the new encryption and a randomly generated secret
2. Alice: toxcore receives the message and does the Initial Steps and saves Bob´s Public Key temporarily
3. Alice: toxcore sends a message to Bob also encrypted with the new Public Key she received of Bob which contains her Public Key for the new encryption algorithm and the hash of the randomly generated secret
4. Bob: toxcore checks whether the hash of the secret matches and saves Alice´s Public Key temporarily
Encryption Check:
5. Bob: toxcore sends a 2nd randomly generated secret to Alice which is additionally encrypted with her previously received Public Key
6. Alice: toxcore decrypts the message and hashes the secret and sends it encrypted with his Public Key to Bob and saves Bob´s Public Key permanently
Alice knows now that Bob has her Public Key
7. Bob: toxcore checks whether the hash of the secret matches and if they do he adds Alice´s Public Key permanently
Bob knows now that Alice has her Public Key

Important:
New Keys have to be exchanged
-->At the time of the Key exchange the connection has to be secure.
New Public Keys have to be visible to the user and should also be implemented into a QR Code mechanism (once it is finished)

My English might not always be perfect because I am no native English speaker.
